edx_django_utils/monitoring/internal/code_owner/middleware.py

```diff
@@ -0,0 +1,147 @@
+        This can be used to find missing mappings.
+
+    """
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        response = self.get_response(request)
+        self._set_code_owner_attribute(request)
+        return response
+
+    def process_exception(self, request, exception):
+        self._set_code_owner_attribute(request)
+
+    def _set_code_owner_attribute(self, request):
+        """
+        Sets the code_owner custom attribute, as well as several supporting custom attributes.
+
+        See CodeOwnerMonitoringMiddleware docstring for a complete list of attributes.
+
+        """
+        code_owner, path_error = self._set_code_owner_attribute_from_path(request)
+        if code_owner:
+            set_custom_attribute('code_owner', code_owner)
+            return
+        if not path_error:
+            # module found, but mapping wasn't configured
+            code_owner = self._set_code_owner_attribute_catch_all()
+            if code_owner:
+                set_custom_attribute('code_owner', code_owner)
+            return
+
+        code_owner, transaction_error = self._set_code_owner_attribute_from_current_transaction(request)
+        if code_owner:
+            set_custom_attribute('code_owner', code_owner)
+            return
+        if not transaction_error:
+            # transaction name found, but mapping wasn't configured
+            code_owner = self._set_code_owner_attribute_catch_all()
+            if code_owner:
+                set_custom_attribute('code_owner', code_owner)
+            return
+
+        code_owner = self._set_code_owner_attribute_catch_all()
+        if code_owner:
+            set_custom_attribute('code_owner', code_owner)
+            return
+
+        # only report errors if code_owner couldn't be found, including catch-all
+        if path_error:
+            set_custom_attribute('code_owner_path_error', path_error)
+        if transaction_error:
+            set_custom_attribute('code_owner_transaction_error', transaction_error)
+
+    def _set_code_owner_attribute_from_path(self, request):
+        """
+        Uses the request path to find the view_func and then sets code owner attributes based on the view.
+
+        Side-effects:
+            Sets code_owner_path_module custom attribute, used to determine code_owner
+
+        Returns:
+            (str, str): (code_owner, error_message), where at least one of these should be None
+
+        """
+        if not is_code_owner_mappings_configured():
+            return None, None
+
+        try:
+            view_func, _, _ = resolve(request.path)
+            path_module = view_func.__module__
+            set_custom_attribute('code_owner_path_module', path_module)
+            code_owner = get_code_owner_from_module(path_module)
+            return code_owner, None
+        except Exception as e:  # pylint: disable=broad-except
```

Comment:
Maybe this is a bad pattern, but it was put in place because I want to ensure that nothing can get through, even unplanned for errors, because I don't want to chance breaking a transaction.  I want to ensure this code is safe.

Reply:
Hey! :wave: Sorry to pitch in, but somehow I noticed this conversation in my inbox.

I think we must really, really avoid this pattern. I see a few major issues here:

1. The exception is cast as a string, which will prevent later raising an exception with the right type and the right error message. For instance, we won't be able to do `logging.exception(...)` anymore (which is very useful, e.g with sentry).
2. The Exception type is overly broad, and this will silence things such as SyntaxError, AttributeError, ImportError -- errors that we do want to learn about, whether it is in development or production. This [happens](https://github.com/edx/course-discovery/pull/2730).

I might not understand the full context here: I thought that the whole point of transactions was that we could interrupt them and be ok, because the entire transaction is rolled back on error?

Reply:
Thanks @regisb.  When this code was written, I very intentionally did **not** want this to interrupt or break any transactions.  At this time, we will be relying on this data more and more, so I should be able to fix this in a future PR when I updated some of the code_owner code.